src/main/rx/ghst.c

```diff
@@ -0,0 +1,312 @@
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "platform.h"
+
+#ifdef USE_SERIALRX_GHST
+
+#include "build/build_config.h"
+#include "build/debug.h"
+
+#include "common/crc.h"
+#include "common/maths.h"
+#include "common/utils.h"
+
+#include "drivers/serial.h"
+#include "drivers/serial_uart.h"
+#include "drivers/system.h"
+#include "drivers/time.h"
+
+#include "io/serial.h"
+
+#include "rx/rx.h"
+#include "rx/ghst.h"
+
+#include "telemetry/ghst.h"
+
+#define GHST_PORT_OPTIONS               (SERIAL_STOPBITS_1 | SERIAL_PARITY_NO | SERIAL_BIDIR | SERIAL_BIDIR_PP)
+#define GHST_PORT_MODE                  MODE_RXTX   // bidirectional on single pin
+
+#define GHST_MAX_FRAME_TIME_US          500         // 14 bytes @ 420k = ~450us
+#define GHST_TIME_BETWEEN_FRAMES_US     4500        // fastest frame rate = 222.22Hz, or 4500us
+
+// define the time window after the end of the last received packet where telemetry packets may be sent
+// NOTE: This allows the Rx to double-up on Rx packets to transmit data other than servo data, but
+// only if sent < 1ms after the servo data packet.
+#define GHST_RX_TO_TELEMETRY_MIN_US     1000
+#define GHST_RX_TO_TELEMETRY_MAX_US     2000
+
+#define GHST_PAYLOAD_OFFSET offsetof(ghstFrameDef_t, type)
+
+STATIC_UNIT_TESTED volatile bool ghstFrameAvailable = false;
+STATIC_UNIT_TESTED volatile bool ghstValidatedFrameAvailable = false;
+STATIC_UNIT_TESTED volatile bool ghstTransmittingTelemetry = false;
+
+STATIC_UNIT_TESTED ghstFrame_t ghstIncomingFrame;   // incoming frame, raw, not CRC checked, destination address not checked
+STATIC_UNIT_TESTED ghstFrame_t ghstValidatedFrame;  // validated frame, CRC is ok, destination address is ok, ready for decode
+
+STATIC_UNIT_TESTED uint32_t ghstChannelData[GHST_MAX_NUM_CHANNELS];
+
+static serialPort_t *serialPort;
+static timeUs_t ghstRxFrameStartAtUs = 0;
+static timeUs_t ghstRxFrameEndAtUs = 0;
+static uint8_t telemetryBuf[GHST_FRAME_SIZE_MAX];
+static uint8_t telemetryBufLen = 0;
+
+/* GHST Protocol
+ * Ghost uses 420k baud single-wire, half duplex connection, connected to a FC UART 'Tx' pin
+ * Each control packet is interleaved with one or more corresponding downlink packets
+ *
+ * Uplink packet format (Control packets)
+ * <Addr><Len><Type><Payload><CRC>
+ *
+ * Addr:        u8      Destination address
+ * Len          u8      Length includes the packet ID, but not the CRC
+ * CRC          u8
+ *
+ * Ghost packets are designed to be as short as possible, for minimum latency.
+ *
+ * Note that the GHST protocol does not handle, itself, failsafe conditions. Packets are passed from
```

Comment:
<rant>This is soooooo wrong. IRC is shifting the blame to the flight controller</rant>

Reply:
So, we're depriving the FC of some information receiver potentially has, depending on the modem chip uses: number of bits corrected by FEC, RSSI of the each received packet, and expect the FC to figure out the signal loss exclusively by calculating and observing the packet loss. Not very well engineered IMO. GHST receiver should either communicate all radio-related data it has to the FC, or better - detect and communicate the failsafe condition to the FC as all other RX systems out there do.